colossalai/inference/core/engine.py

```diff
@@ -1,65 +1,196 @@
     """
 
     def __init__(
         self,
-        tokenizer: str = None,
         inference_config: Optional["InferenceConfig"] = None,
         verbose: bool = False,
     ) -> None:
         assert inference_config, "Please provide inference_config."
 
-        self._init_model()
-        # cache_config may need to be modified later.
-        # self.request_handler = RequestHandler(cache_config)
-        self.tokenizer = tokenizer
-        self.hf_model_config = AutoConfig.from_pretrained(
-            self.model, trust_remote_code=self.trust_remote_code, revision=self.revision
+        self.tokenizer = get_tokenizer(
+            inference_config.tokenizer,
+            use_fast_tokenizer=inference_config.use_fast_tokenizer,
+            trust_remote_code=inference_config.trust_remote_code,
         )
+        self.tokenizer.pad_token = self.tokenizer.eos_token
+        self.inference_config = inference_config
+
+        self.verbose = verbose
         if verbose:
-            self.logger = Logger()
+            self.logger = get_dist_logger(__name__)
 
-    def _init_model(self):
+        self._init_model_and_hf_config()
+        self.cache_manager = KVCacheManager(self.inference_config, self.hf_model_config, verbose)
+        self.requset_handler = RequestHandler(self.inference_config)
+
+        self.counter = count()
+        self._verify_config()
+
+    def _init_model_and_hf_config(self):
         """
-        Initialize model and distributed training environment(if needed).
-        May need to provide two different initialization methods:
-            1. 用户自定义(from local path)
-            2. 从checkpoint加载(hugging face)
+        Initialize model.
         """
+        if isinstance(self.inference_config.model, str):
+            self.model = init_model(self.inference_config, self.hf_model_config)
+            self.hf_model_config = AutoConfig.from_pretrained(
+                self.inference_config.model,
+                trust_remote_code=self.inference_config.trust_remote_code,
+                revision=self.inference_config.revision,
+            )
+        elif isinstance(self.inference_config.model, nn.Module):
+            self.model = self.inference_config.model
+            self.hf_model_config = self.model.config
+        else:
+            raise ValueError(
+                f"The type of inference_config.model should be str or nn.Module, but get {type(self.inference_config.model)}"
+            )
+
+        if self.verbose:
+            self.logger.info("Start to initialize model")
+
+        if isinstance(self.inference_config.model, str):
+            self.model = init_model(self.inference_config, self.hf_model_config)
+            self.hf_model_config = AutoConfig.from_pretrained(
+                self.inference_config.model,
+                trust_remote_code=self.inference_config.trust_remote_code,
+                revision=self.inference_config.revision,
+            )
```

Comment:
This kind of confuses me. There is a dedicated function for model and tokenizer creation, but the config creation is embedded into a member method.

Reply:
Because we directly use  model passed by users, we will use the "self.model_config = model.config" to obtain the model config.